identity/aziot-identityd/src/renewal.rs

```diff
@@ -0,0 +1,301 @@
+        let (key_id, key_handle) = if self.rotate_key {
+            let key_id = format!("{}-temp", key_id);
+
+            if let Ok(key_handle) = self.key_client.load_key_pair(&key_id).await {
+                self.key_client
+                    .delete_key_pair(&key_handle)
+                    .await
+                    .map_err(|_| {
+                        cert_renewal::Error::retryable_error("failed to clear temp key")
+                    })?;
+            }
+
+            let key_handle = self
+                .key_client
+                .create_key_pair_if_not_exists(&key_id, Some("rsa-2048:*"))
+                .await
+                .map_err(|_| cert_renewal::Error::retryable_error("failed to generate temp key"))?;
+
+            (key_id, key_handle)
+        } else {
+            let key_handle = self.key_client.load_key_pair(key_id).await.map_err(|_| {
+                cert_renewal::Error::retryable_error("failed to get identity cert key")
+            })?;
+
+            (key_id.to_string(), key_handle)
+        };
+
+        let (private_key, public_key) = crate::get_keys(key_handle, &self.key_engine)
+            .await
+            .map_err(cert_renewal::Error::retryable_error)?;
+
+        // Determine the subject of the old cert. This will be the subject of the new cert.
+        let subject = if let Some(subject) = old_cert
+            .subject_name()
+            .entries_by_nid(openssl::nid::Nid::COMMONNAME)
+            .next()
+        {
+            String::from_utf8(subject.data().as_slice().into())
+                .map_err(|_| cert_renewal::Error::fatal_error("bad cert subject"))?
+        } else {
+            return Err(cert_renewal::Error::fatal_error(
+                "cannot determine subject for csr",
+            ));
+        };
+
+        // Generate a CSR and issue the new cert under a temporary cert ID. The temporary ID
+        // does not need to be persisted, so delete it after the cert is succesfully created.
+        let csr = crate::create_csr(&subject, &public_key, &private_key, None)
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to create csr"))?;
+
+        let new_cert = self
+            .cert_client
+            .create_cert(&self.temp_cert, &csr, None)
+            .await
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to create new cert"))?;
+
+        if let Err(err) = self.cert_client.delete_cert(&self.temp_cert).await {
+            log::warn!(
+                "Failed to delete temporary certificate created by cert renewal: {}",
+                err
+            );
+        }
+
+        let new_cert = openssl::x509::X509::from_pem(&new_cert)
+            .map_err(|_| cert_renewal::Error::retryable_error("failed to parse new cert"))?;
+
+        Ok((new_cert, key_id))
+    }
+
+    async fn write_credentials(
+        &mut self,
+        _old_cert: &openssl::x509::X509,
```

Comment:
Remove unused param?

Reply:
It's used by other implementers of this trait. They revert to using the old cert if something fails. Renewal of the device identity certificate doesn't use it because it has already reprovisioned the device with the new cert, so it would have to provision again to revert back.